es-core/src/AudioManager.cpp

```diff
@@ -135,21 +112,169 @@ void AudioManager::unregisterSound(std::shared_ptr<Sound> & sound)
 		}
+	}		
+}
+
+void AudioManager::playRandomMusic(bool continueIfPlaying)
+{
+	if (!Settings::getInstance()->getBool("EnableMusic"))
+		return;
+
+	std::vector<std::string> musics;
+	
+	// check in RetroPie music directory
+	if (musics.empty())
+		getMusicIn(Utils::FileSystem::getHomePath() + "/RetroPie/roms/music", musics);
+  
+	// check in system sound directory
+	if (musics.empty())
+		getMusicIn("/usr/share/RetroPie/music", musics);
+  
+	// check in .emulationstation/music directory
+	if (musics.empty())
+		getMusicIn(Utils::FileSystem::getHomePath() + "/.emulationstation/music", musics);
+
+	if (musics.empty())
+		return;
+
+#if defined(WIN32)
+	srand(time(NULL) % getpid());
+#else
+	srand(time(NULL) % getpid() + getppid());
+#endif
+
+	int randomIndex = rand() % musics.size();
+
+	// continue playing ?
+	if (mCurrentMusic != nullptr && continueIfPlaying)
+		return;
+
+	playMusic(musics.at(randomIndex));
+}
+
+void AudioManager::playMusic(std::string path)
+{
+	if (!mInitialized)
+		return;
+
+	// free the previous music
+	stopMusic(false);
+
+	if (!Settings::getInstance()->getBool("EnableMusic"))
+		return;
+
+	// load a new music
+	mCurrentMusic = Mix_LoadMUS(path.c_str());
+	if (mCurrentMusic == NULL)
+	{
+		LOG(LogError) << Mix_GetError() << " for " << path;
+		return;
 	}
-	//pause audio
-	SDL_PauseAudio(1);
+
+	if (Mix_FadeInMusic(mCurrentMusic, 1, 1000) == -1)
+	{
+		stopMusic();
+		return;
+	}
+
+	Mix_HookMusicFinished(AudioManager::musicEnd_callback);
+}
+
+void AudioManager::musicEnd_callback()
+{
+    if (!Settings::getInstance()->getBool("EnableMusic"))
+		return;
+	else
+		AudioManager::getInstance()->playRandomMusic(false);
+}
+
+void AudioManager::stopMusic(bool fadeOut)
+{
+	if (mCurrentMusic == NULL)
+		return;
+
+	Mix_HookMusicFinished(nullptr);
+
+	if (fadeOut)
+	{
+		while (!Mix_FadeOutMusic(2000) && Mix_PlayingMusic())
+			SDL_Delay(100);
+	}
+
+	Mix_FreeMusic(mCurrentMusic);
+	mCurrentMusic = NULL;
+}
+
+int AudioManager::getMaxMusicVolume()
+{
+	int ret = (Settings::getInstance()->getInt("MusicVolume") * MIX_MAX_VOLUME) / 100;
+	if (ret > MIX_MAX_VOLUME)
+		return MIX_MAX_VOLUME;
+
+	if (ret < 0)
+		return 0;
+
+	return ret;
+}
+
+void AudioManager::update(int deltaTime)
+{
+	if (sInstance == nullptr || !sInstance->mInitialized || !Settings::getInstance()->getBool("EnableMusic"))
+		return;
+
+	float deltaVol = deltaTime / 8.0f;
+
+//	#define MINVOL 5
+
+	int maxVol = getMaxMusicVolume();
+	int minVol = maxVol / 20;
+	if (maxVol > 0 && minVol == 0)
+		minVol = 1;
+
+	/*if (sInstance->mMusicVolume > minVol)
+	{
+		sInstance->mMusicVolume -= deltaVol;
+		if (sInstance->mMusicVolume < minVol)
+			sInstance->mMusicVolume = minVol;
+
+		Mix_VolumeMusic((int)sInstance->mMusicVolume);
+	}
+	else if (sInstance->mMusicVolume < maxVol)
+	{
+		sInstance->mMusicVolume += deltaVol;
+		if (sInstance->mMusicVolume > maxVol)
+			sInstance->mMusicVolume = maxVol;
+
+		Mix_VolumeMusic((int)sInstance->mMusicVolume);
+	}*/
```

Comment:
Commented out code needed?

Reply:
Added comments to why it's commented out. This is part of another feature that could be added later on.